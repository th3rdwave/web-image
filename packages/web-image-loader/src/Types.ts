export interface ResolvedImageSource {
  /**
   * Density scale for this image (1x, 2x, 3x)
   */
  scale: number;
  /**
   * Final uri for the image.
   */
  uri: string;
  /**
   * Mime type
   */
  type: string;
}

export interface WebpackResolvedImage {
  outputPath: string;
  publicPath: string;
  /**
   * Density scale for this image (1x, 2x, 3x)
   */
  scale: number;
  /**
   * Mime type
   */
  type: string;
}

export interface AdaptativeImageSource {
  /**
   * srcset string (https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attr-srcset)
   */
  srcSet: string;
  /**
   * Mime type
   */
  type: string;
}

export interface AdaptativeImageData {
  width: number;
  height: number;
  uri: string;
  sources?: AdaptativeImageSource[];
}
