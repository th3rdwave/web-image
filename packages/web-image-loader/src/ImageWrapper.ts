import groupBy from 'lodash/groupBy';
import { WebpackResolvedImage } from './Types';

const SORT = ['image/webp', 'image/avif'];

export function createImageWrapper(classPath: string, esModule: boolean) {
  return (
    size: { width: number; height: number },
    images: WebpackResolvedImage[],
  ): string => {
    const imagesByType = groupBy(images, (img) => img.type);
    // Template strings are a bit weird here but this is the price
    // to pay to make the generated code beautiful :S
    const sources = `[
    ${Object.values(imagesByType)
      // Make sure avif then webp comes first.
      .sort((a, b) => {
        const ia = SORT.indexOf(a[0].type);
        const ib = SORT.indexOf(b[0].type);
        return ib - ia;
      })
      .map(
        (group) => `{
      srcSet: ${group
        .map((img) => `${img.publicPath} + " ${img.scale}x"`)
        .join(' + "," +\n        ')},
      type: "${group[0].type}"
    }`,
      )
      .join(',\n    ')}
  ]`;

    return `${
      esModule
        ? `import {AdaptiveImage} from ${classPath}`
        : `var AdaptiveImage = require(${classPath}).AdaptiveImage`
    };

${esModule ? 'export default' : 'module.exports ='} new AdaptiveImage({
  uri: ${images[0].publicPath},
  width: ${size.width},
  height: ${size.height},
  sources: ${sources}
});
`;
  };
}
