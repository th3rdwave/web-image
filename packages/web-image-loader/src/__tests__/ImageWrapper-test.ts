import { parse } from '@babel/parser';
import { createImageWrapper } from '../ImageWrapper';
import { WebpackResolvedImage } from '../Types';

const IMAGE_WRAPPER_PATH =
  '"../../node_modules/@th3rdwave/web-image-loader/lib/commonjs/AdaptiveImage"';

const IMAGE_SIZE = { width: 100, height: 100 };

const createImages = (image: string): WebpackResolvedImage[] =>
  Array.from({ length: 3 }).map((_, i) => {
    const [name, ext] = image.split('.');
    return {
      publicPath: `__webpack_public_path__ + "static/${name}@${i + 1}x.${ext}"`,
      outputPath: image,
      scale: i + 1,
      type: `image/${ext}`,
    };
  });

describe('ImageWrapper', () => {
  it('generates valid javascript', () => {
    const imageWrapper = createImageWrapper(IMAGE_WRAPPER_PATH, true);
    const code = imageWrapper(IMAGE_SIZE, [
      ...createImages('img.png'),
      ...createImages('img.webp'),
    ]);
    expect(parse(code, { sourceType: 'module' })).toEqual(expect.any(Object));
  });

  it('generates es modules source', () => {
    const imageWrapper = createImageWrapper(IMAGE_WRAPPER_PATH, true);
    expect(imageWrapper(IMAGE_SIZE, createImages('img.png'))).toMatchSnapshot();
  });

  it('generates commonjs source', () => {
    const imageWrapper = createImageWrapper(IMAGE_WRAPPER_PATH, false);
    expect(imageWrapper(IMAGE_SIZE, createImages('img.png'))).toMatchSnapshot();
  });

  it('handles multiple sources', () => {
    const imageWrapper = createImageWrapper(IMAGE_WRAPPER_PATH, true);
    expect(
      imageWrapper(IMAGE_SIZE, [
        ...createImages('img.png'),
        ...createImages('img.avif'),
        ...createImages('img.webp'),
      ]),
    ).toMatchSnapshot();
  });
});
