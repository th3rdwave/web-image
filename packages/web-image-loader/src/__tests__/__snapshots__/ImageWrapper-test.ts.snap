// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageWrapper generates commonjs source 1`] = `
"var AdaptiveImage = require(\\"../../node_modules/@th3rdwave/web-image-loader/lib/commonjs/AdaptiveImage\\").AdaptiveImage;

module.exports = new AdaptiveImage({
  uri: __webpack_public_path__ + \\"static/img@1x.png\\",
  width: 100,
  height: 100,
  sources: [
    {
      srcSet: __webpack_public_path__ + \\"static/img@1x.png\\" + \\" 1x\\" + \\",\\" +
        __webpack_public_path__ + \\"static/img@2x.png\\" + \\" 2x\\" + \\",\\" +
        __webpack_public_path__ + \\"static/img@3x.png\\" + \\" 3x\\",
      type: \\"image/png\\"
    }
  ]
});
"
`;

exports[`ImageWrapper generates es modules source 1`] = `
"import {AdaptiveImage} from \\"../../node_modules/@th3rdwave/web-image-loader/lib/commonjs/AdaptiveImage\\";

export default new AdaptiveImage({
  uri: __webpack_public_path__ + \\"static/img@1x.png\\",
  width: 100,
  height: 100,
  sources: [
    {
      srcSet: __webpack_public_path__ + \\"static/img@1x.png\\" + \\" 1x\\" + \\",\\" +
        __webpack_public_path__ + \\"static/img@2x.png\\" + \\" 2x\\" + \\",\\" +
        __webpack_public_path__ + \\"static/img@3x.png\\" + \\" 3x\\",
      type: \\"image/png\\"
    }
  ]
});
"
`;

exports[`ImageWrapper handles multiple sources 1`] = `
"import {AdaptiveImage} from \\"../../node_modules/@th3rdwave/web-image-loader/lib/commonjs/AdaptiveImage\\";

export default new AdaptiveImage({
  uri: __webpack_public_path__ + \\"static/img@1x.png\\",
  width: 100,
  height: 100,
  sources: [
    {
      srcSet: __webpack_public_path__ + \\"static/img@1x.avif\\" + \\" 1x\\" + \\",\\" +
        __webpack_public_path__ + \\"static/img@2x.avif\\" + \\" 2x\\" + \\",\\" +
        __webpack_public_path__ + \\"static/img@3x.avif\\" + \\" 3x\\",
      type: \\"image/avif\\"
    },
    {
      srcSet: __webpack_public_path__ + \\"static/img@1x.webp\\" + \\" 1x\\" + \\",\\" +
        __webpack_public_path__ + \\"static/img@2x.webp\\" + \\" 2x\\" + \\",\\" +
        __webpack_public_path__ + \\"static/img@3x.webp\\" + \\" 3x\\",
      type: \\"image/webp\\"
    },
    {
      srcSet: __webpack_public_path__ + \\"static/img@1x.png\\" + \\" 1x\\" + \\",\\" +
        __webpack_public_path__ + \\"static/img@2x.png\\" + \\" 2x\\" + \\",\\" +
        __webpack_public_path__ + \\"static/img@3x.png\\" + \\" 3x\\",
      type: \\"image/png\\"
    }
  ]
});
"
`;
